etcd/v1alpha1/types_pacemakercluster.go

```diff
@@ -0,0 +1,436 @@
+	//   to cover states like init, wait_for_ping, starting_daemons, shutting_down, shutdown_complete, etc.
+	// +optional
+	PacemakerDaemonState PacemakerDaemonStateType `json:"pacemakerDaemonState,omitempty"`
+
+	// quorumStatus indicates if the cluster has quorum
+	// QuorumStatusType can be one of the following values:
+	// - Quorate - the cluster has quorum
+	// - NoQuorum - the cluster does not have quorum
+	// +optional
+	QuorumStatus QuorumStatusType `json:"quorumStatus,omitempty"`
+
+	// nodesOnline is the count of online nodes
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=2
+	// +optional
+	NodesOnline *int32 `json:"nodesOnline,omitempty"`
+
+	// nodesTotal is the total count of configured nodes
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=2
+	// +optional
+	NodesTotal *int32 `json:"nodesTotal,omitempty"`
+
+	// resourcesStarted is the count of started resources
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=16
+	// +optional
+	ResourcesStarted *int32 `json:"resourcesStarted,omitempty"`
+
+	// resourcesTotal is the total count of configured resources
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=16
+	// +optional
+	ResourcesTotal *int32 `json:"resourcesTotal,omitempty"`
+}
+
+// NodeStatus represents the status of a single node in the Pacemaker cluster
+type PacemakerNodeStatus struct {
+	// name is the name of the node
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=256
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// ipv4Address is the IPv4 address of the node, if registered via IPv4
+	// +kubebuilder:validation:MinLength=7
+	// +kubebuilder:validation:MaxLength=15
+	// +kubebuilder:validation:Pattern="^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$"
+	// +optional
+	IPv4Address string `json:"ipv4Address,omitempty"`
+
+	// ipv6Address is the IPv6 address of the node, if registered via IPv6
+	// +kubebuilder:validation:MinLength=2
+	// +kubebuilder:validation:MaxLength=39
+	// +kubebuilder:validation:Format=ipv6
+	// +kubebuilder:validation:Pattern=`^(([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))$`
+	// +optional
+	IPv6Address string `json:"ipv6Address,omitempty"`
```

Comment:
CEL has IP validations that support both IPv4 and IPv6. It would be better to combine them and use the CEL validations instead, sorry for the back and forth here.

Reply:
In terms of why isIP() is not sufficient:
It depends how strict we want to be in this API. The IPs we use are expected to identify the nodes as their endpoint-identities for etcd. So they should be unique, they should (ideally) be in their canonical form, and they should be the kinds of IPs that are not reserved for special cases.

I defaulted to more strict because I've never written one of these so I decided to err on the side of adding the restrictions that made sense to me.

Reply:
Turns out enforcing !isGlobalUnicast() was also eliminating all the IPs I wanted to accept. :sweat_smile: